### ISSUE TITLE
Bias gradient support?

### ISSUE THREAD

**Issue by ardagoreci** [pytorch-labs/attention-gym#20]
I was wondering if there are any plans to support gradient flow back to biases that are added to the score function. For instance, if I add a scalar to the: score + b where b is coming from a learnable projection, the gradients should flow back to those layers. 

Currently, I get the following error when I try this:

AssertionError: Captured buffers that require grad are not yet supported.

This feature would be very useful for the attention mechanisms in AlphaFold for computational biology (a lot of pair biasing in those attention components).

**Comment by drisspg** [comment]
Makes sense, we very much plan to support this however there is some complexity on how you generate the bwd for this because broadcasts in the forward become reductions in the backward. We are discussing the implementation now and I will keep you posted

**Comment by ardagoreci** [comment]
Sounds great, thank you!

**Comment by eballesteros** [comment]
+1 to this request! It would be very useful for T5-style learned relative positional biases

```python
def learned_relative_positional(score, b, h, q_idx, kv_idx):
    r_pos = (q_idx - kv_idx)
    return score + pos_emb[r_pos]

```

**Comment by ardagoreci** [comment]
Another update, we recently released a preliminary open source implementation of AlphaFold3 here: https://github.com/Ligo-Biosciences/AlphaFold3.  The model is still not nearly scalable as it could be and attention ops are memory bound for the most part. I can migrate to FlexAttention as soon as there is bias gradient support!

**Comment by ds-kczerski** [comment]
Another +1!

I would be interested to have a support for a function like that:
```
def wrapped_agraph(distances: Tensor, distance_encoders: Embedding):
    """
    wrapping score modification function for flex
    
    distances - torch tensor [batch_size, seq_len, seq_len], example: torch.Size([32, 49, 49], int matrix
    distance_encoders - Embedding layer [embeddings_count, heads] (see: from torch.nn import Embedding), example: Embedding(13, 16)
    """
    def agraph(score, b, h, q_idx, kv_idx):
        idx_d = distances[b, q_idx, kv_idx]
        return score + distance_encoders(idx_d)[h]
    return agraph
```

**Comment by ViktorooReps** [comment]
Another +1

**Comment by drisspg** [comment]
We are actively working on this now, will post back here with further updates

**Comment by pengzhenghao** [comment]
+1 see #69

**Comment by drisspg** [comment]
Will have updates soon :)

**Comment by pengzhenghao** [comment]
Thanks for the update! Will try out when available. Do you have a release timeline? @drisspg

**Comment by bkoszarsky** [comment]
interested as well

**Comment by drisspg** [comment]
We have a working implementation in the nightlies (PR: https://github.com/pytorch/pytorch/pull/137452).

Earlier, we asked for feedback on a particular trade-off we needed to make to implement this: speed vs. extra memory. People said they would sacrifice some speed to avoid having to realize a full `(Batch, N_Heads, Seq_len, Seq_len)`attention score.

This is what we have implemented in the above PR. It will produce a grad_bias that is the same shape and utilizes atomic writes to accumulate broadcasted gradients from the forward pass. This does, however, incur a non-zero performance penalty. I have seen this range from 2x slower backward pass to ~10x slower (in a pretty adversarial case). 
I am currently profiling and categorizing the performance characteristics of the solution and working with some Triton folks to find ways to optimize. Another caveat is that it's non-deterministic.

TLDR: You can kick the tires today if you install nightlies.

I would love for people to provide real-world use cases I can use for testing. Feel free to comment on this issue with real-world examples I can measure performance on. 

I'm excited to see what people build 😊!

**Comment by pengzhenghao** [comment]
Thanks for the update! I can share an implementation with graph network:  https://github.com/rainmaker22/SMART/blob/main/smart/layers/attention_layer.py

The project is for predicting multi-agent behaviors in driving scenarios. We need to relative relations between agents and have to involve per-Q-K relations into the attention scores. The above implementation actually perform quite well. But it doesn't use any flash/flex attention to accelerate the scaled dot product.

**Comment by lhallee** [comment]
Has anyone used the nightly version for T5 relative attention bias yet?

**Comment by hutslib** [comment]
@drisspg 
Hi, I have just tried the nightly version and I came across two problems. 
```
# ERROR1: If I run compile then I get an error
# flex_atten =  torch.compile(flex_attention)
# y = flex_atten(query=q, key=k, value=v, score_mod=rel_pe_score)
# instead I can use the following code, the version without a compile
y = flex_attention(query=q, key=k, value=v, score_mod=rel_pe_score)
# ERROR2： When testing the nightly version, it can run successfully with trainable bias, however,
the GPU memory usage becomes larger and then catches an oom error. A version without using flex attention can run successfully (no oom error)
```

**Comment by drisspg** [comment]
@hutslib yeah I think w/ the above description of your problem it will be hard to debug. If you are able to create a more minimal repro I can look at or share some code pointers with me that would real help with debugging

**Comment by hutslib** [comment]
@drisspg Hi, 
I provided a simplified version of the code to replicate the workflow. The code captures an error; however, I am unsure if it is the same error encountered when running my entire code, as the output is more complex when executing the entire workflow. Could you help to first analyze this?

**Problem for the provided code:**
If we adjust the bs, T and topk value, sometimes the code will raise an error. 
Reduce the bs, T, topk value, somtimes avoid the error.
`RuntimeError: CUDA error: an illegal memory access was encountered`

BTW, I also need some suggestion for select kernel_options.

Code sample 
https://github.com/hutslib/hutslib/blob/main/text-flex_attention.py

**Comment by timt51** [comment]
@hutslib fyi that link may not be accessible; it gives me 404 not found

**Comment by hutslib** [comment]
> [@drisspg](https://github.com/drisspg) Hi, I provided a simplified version of the code to replicate the workflow. Code sample https://github.com/hutslib/hutslib/blob/main/test-flex_attention.py Could you help to analyze this?
> 
> **Problem for the provided code:** If we adjust the bs, T and topk value, sometimes the code will raise an error. Reduce the bs, T, topk value, somtimes avoid the error. `RuntimeError: CUDA error: an illegal memory access was encountered`
> 
> BTW, I also need some suggestion for select kernel_options.

Change the visibility of the repo, and now the link is accessible.

**Comment by lhallee** [comment]
Any update on this @drisspg?

**Comment by drisspg** [comment]
We added support in 2.7.0 for learnable biases to PyTorch, there are still some edges but it should work. Do you have an implementation that is not working correctly that repros?

**Comment by ArthurSBianchessi** [comment]
(hi, I am with @lucaskup )

It took some time to pinpoint the problem, since the problem was a little dificult to reproduce without knowing the problem.
Bellow is a implementation of [Scalable-Softmax](https://arxiv.org/abs/2501.19399), which does suffer from the same problem.

```python
import torch
from torch import nn
from torch.nn.attention.flex_attention import flex_attention

batch_size = 2
n_heads = 8
seq_length = 1024
h_dim = 64
DEVICE = 'cuda:0'

class AttentionSSMax(nn.Module):
    def __init__(self):
        super(AttentionSSMax, self).__init__()
        self.scale = nn.Parameter(torch.ones((n_heads, 1), dtype=torch.float))

    def forward(self, q, k, v):
        pos_mul = torch.arange(1, q.size(-2) + 1, device=q.device).log() * self.scale
        ########################################################################################
        print('Uncomment for the grad to work') 
        ########################################################################################
        def score_mod(score, b, h, q_idx, kv_idx):
            return score * pos_mul[h, q_idx]
        score = flex_attention(q, k, v, score_mod=score_mod)
        return score
        
model = AttentionSSMax()
model.to(DEVICE)
model = torch.compile(model, dynamic=True)


query   = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)
key     = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)
value   = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)

loss = model(query, key, value).sum()
loss.backward()
print(model.scale.grad)
```

When using a parameter such as `scale` directly, it sets the gradient normally, but when calculating an itermediary value, the backward pass ceases working correctly (prints `None`). The issue could be due a synchronization issue since by adding a simple print, it works as intended. It may also be to some diferent code generation by the compiler. This is corraborated by adding any other torch computation instead of the `print()`, which doesn't prevent the problem from occurring.

**Comment by ArthurSBianchessi** [comment]
Any updates?

**Comment by drisspg** [comment]
@ArthurSBianchessi
I can reproduce, still investigating

**Comment by 9race** [comment]
Is this addressed here? https://pytorch.org/blog/flexattention-for-inference/
I see it mentions addition of trainable parameters in the `score_mod` function, but I haven't looked super closely yet.

**Comment by 9race** [comment]
@drisspg  Hi, I've been getting some different errors with torch.compile and learnable biases. I am trying to use FlexAttention with learnable biases (which I've noticed only works with `torch >= 2.7.0` so I can't downgrade to `2.5.1` as suggested here [NoValidChoices](https://github.com/pytorch/torchtune/issues/2392)), and these are the different cases: 

1.  Run without torch.compile. This runs fine, but there is no performance speedup. 

2. Use torch.compile WITHOUT the added trainable biases --> `self.flex_attention = torch.compile(flex_attention, dynamic=False, mode="max-autotune")`. I want to do this in order to get a performance speedup. However, this gives a memory error `torch._inductor.exc.InductorError: RuntimeError: No valid triton configs. OutOfResources: out of resource: shared memory, Required: 180480, Hardware limit: 166912. Reducing block sizes or `num_stages` may help.`. Even with a batch size of 1, I get the same error. The model at least runs though despite these errors constantly popping up. 

3. Use torch.compile WITH trainable biases added in `score_mod`. I get different errors depending on the mode. With `max-autotune`, I get
`[rank0]: torch._inductor.exc.InductorError: LoweringException: NoValidChoicesError:
[rank0]:   target: flex_attention`

However, with the default mode, I get memory errors in 2, and in both cases it doesn't run.

Any assistance on using torch.compile with learnable biases?

**Comment by ArthurSBianchessi** [comment]
> (hi, I am with [@lucaskup](https://github.com/lucaskup) )
> 
> It took some time to pinpoint the problem, since it was a little dificult to reproduce due to its erratic behavior. Bellow is a implementation of [Scalable-Softmax](https://arxiv.org/abs/2501.19399), which does suffer from the same problem.
> 
> import torch
> from torch import nn
> from torch.nn.attention.flex_attention import flex_attention
> 
> batch_size = 2
> n_heads = 8
> seq_length = 1024
> h_dim = 64
> DEVICE = 'cuda:0'
> 
> class AttentionSSMax(nn.Module):
>     def __init__(self):
>         super(AttentionSSMax, self).__init__()
>         self.scale = nn.Parameter(torch.ones((n_heads, 1), dtype=torch.float))
> 
>     def forward(self, q, k, v):
>         pos_mul = torch.arange(1, q.size(-2) + 1, device=q.device).log() * self.scale
>         ########################################################################################
>         print('Uncomment for the grad to work') 
>         ########################################################################################
>         def score_mod(score, b, h, q_idx, kv_idx):
>             return score * pos_mul[h, q_idx]
>         score = flex_attention(q, k, v, score_mod=score_mod)
>         return score
>         
> model = AttentionSSMax()
> model.to(DEVICE)
> model = torch.compile(model, dynamic=True)
> 
> 
> query   = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)
> key     = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)
> value   = torch.randn(batch_size, n_heads, seq_length, h_dim, device=DEVICE, requires_grad=True)
> 
> loss = model(query, key, value).sum()
> loss.backward()
> print(model.scale.grad)
> 
> When using a parameter such as `scale` directly, it sets the gradient normally, but when calculating an itermediary value, the backward pass ceases working correctly (prints `None`). The issue could be due a synchronization issue, since by adding a simple print, it works as intended. It may also be to some diferent code generation by the compiler. This is corraborated by adding any other torch computation instead of the `print()`, which doesn't prevent the problem from occurring.


Fixed on [PyTorch 2.9](https://github.com/pytorch/pytorch/releases/tag/v2.9.0).
Thanks @drisspg!

**Comment by drisspg** [comment]
Closing as this is has since landed, if you have other issues feel free to open standalone issues
